src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -1047,17 +1120,17 @@ private BoundStatement MakePositionIncrement(CSharpSyntaxNode syntax, BoundLocal
                         type: intType)));
         }
 
-        private void InstrumentForEachStatementCollectionVarDeclaration(BoundForEachStatement original, [NotNullIfNotNull(nameof(collectionVarDecl))] ref BoundStatement? collectionVarDecl)
+        private void InstrumentForEachStatementCollectionVarDeclaration(BoundNode node, [NotNullIfNotNull(nameof(collectionVarDecl))] ref BoundStatement? collectionVarDecl)
         {
-            if (this.Instrument)
+            if (this.Instrument && node is BoundForEachStatement original)
```

Comment:
Consider asserting which types of nodes are expected here and in similar places below.